app/src/main/java/hacking/to/the/gate/CollisionEngine.java

```diff
@@ -48,31 +100,41 @@ private void doCollision(Hittable h1, Hittable h2){
         h2.onCollision(h1);
     }
 
-    public void tick(){
-        for(Iterator<PowerUp> i = mPowerups.iterator();i.hasNext();) {
-            PowerUp p = i.next();
-            if(!mPlayer.isDead()&&p.isVisible()&& detectCollision(mPlayer,p)){
-                doCollision(mPlayer,p);
-            }
+    private void traverse(Hittable root1,Hittable root2){
+
+        traverseHelper(root1,root2);
+        if(root2.getHittableChildren() == null)
+            return;
+        for(Hittable h :root2.getHittableChildren()){
+            traverse(root1,h);
         }
+    }
+    private void traverseHelper(Hittable h1,Hittable h2){
 
-        for(EnemyJet jet:mEnemies){
-            for (Iterator<Bullet> it = jet.getBullets().iterator(); it.hasNext(); ) {
-                Bullet b = it.next();
-                if (!mPlayer.isDead() && detectCollision(mPlayer, b)) {
-                    Log.d("collision", "hit by enemy's bullet");
-                    doCollision(mPlayer,b);
-                }
-            }
-            for(Iterator<Bullet> it = mPlayer.getBullets().iterator(); it.hasNext();){
-                Bullet b = it.next();
-                if(!jet.isDead()&& detectCollision(jet,b)) {
-                    Log.d("collision","hit by player's bullet");
-                    doCollision(jet,b);
+        if(detectCollision(h1,h2)){
+            doCollision(h1,h2);
+        }
+       if(h1.getHittableChildren() == null){
+           return;
+        }
+        for(Hittable h:h1.getHittableChildren()){
+            traverseHelper(h,h2);
+        }
+
+    }
+    public void tick(){
+        mNegativeOnes.clear();
+        mPositiveOnes.clear();
+        mNegativeOnes.addAll(mEnemies);
+        mNegativeOnes.addAll(mPowerups);
+        mPositiveOnes.addAll(mBombs);
+        mPositiveOnes.add(mPlayer);
+        if(mNegativeOnes!= null && mPositiveOnes != null){
+            for(Hittable h : mNegativeOnes){
+                for(Hittable h2 : mPositiveOnes){
+                    traverse(h,h2);
```

Comment:
In theory you can make mNegativeOnes and mPositiveOnes to be the children Hittable of some super Hittable, to avoid loops and make code beautiful. You can try this if you have time. @Jelly-Yu